[feat] Allow setting test variables from the command line (#2117)
eth-cscs/reframe: docs/manpage.rst

.. option:: -S, --setvar=[TEST.]VAR=VAL

   Set variable ``VAR`` in all tests or optionally only in test ``TEST`` to ``VAL``.

   Multiple variables can be set at the same time by passing this option multiple times.
   This option *cannot* change arbitrary test attributes, but only test variables declared with the :attr:`~reframe.core.pipeline.RegressionMixin.variable` built-in.
   If an attempt is made to change an inexistent variable or a test parameter, a warning will be issued.

   ReFrame will try to convert ``VAL`` to the type of the variable.
   If it does not succeed, a warning will be issued and the variable will not be set.
   ``VAL`` can take the special value ``@none`` to denote that the variable must be set to :obj:`None`.

   Sequence and mapping types can also be set from the command line by using the following syntax:

   - Sequence types: ``-S seqvar=1,2,3,4``
   - Mapping types: ``-S mapvar=a:1,b:2,c:3``

   Conversions to arbitrary objects are also supported.
   See :class:`~reframe.utility.typecheck.ConvertibleType` for more details.

   The optional ``TEST.`` prefix refers to the test class name (*not* the test name).

>    If the test name is not explicitly set through the :attr:`~reframe.core.pipeline.RegressionTest.name` attribute or if the test is not parametrised, then the class name is also the test name.

Author: I am trying to see it from a user perspective that may not be so familiar with ReFrame tests. In such cases, it can be confusing for them the "test class name" vs the "test name," given also that tests are listed by name. I could make it the sentence a bit less specific here by saying sth like "generally, the test name is the same as the test class name" and refer to the `name` attribute, which btw needs to be expanded, because we don't mention anything about the default name.

>    If the test name is not explicitly set through the :attr:`~reframe.core.pipeline.RegressionTest.name` attribute or if the test is not parametrised, then the class name is also the test name.

Reviewer: My point is that there is no need for a reference to the `name` attribute at all. The first sentence says that `TEST` is the  class name, not the test name. And the second one explains the default values of the test `name` attribute, even though the previous sentence said that this is attribute has nothing to do with this command line option.

To remove the ambiguity, I'd simply change this to  _The optional ``TEST.`` prefix refers to the test class._

>    If the test name is not explicitly set through the :attr:`~reframe.core.pipeline.RegressionTest.name` attribute or if the test is not parametrised, then the class name is also the test name.

Author: Ok, I removed completely the sentence with the reference to the `name` attribute.

   Variable assignments passed from the command line happen *before* the test is instantiated and is the exact equivalent of assigning a new value to the variable *at the end* of the test class body.
   This has a number of implications that users of this feature should be aware of:

   - In the following test, :attr:`num_tasks` will have always the value ``1`` regardless of any command-line assignment of the variable :attr:`foo`:

   .. code-block:: python

      @rfm.simple_test
      class my_test(rfm.RegressionTest):
          foo = variable(int, value=1)
          num_tasks = foo

>    - If the variable is set in any pipeline hook, the command line assignment will have an effect until the variable assignment in the pipeline hook is reached.

Reviewer 2: ```suggestion
   - If the variable is subsequently set in a pipeline hook, the command line assignment will have an effect only until the variable assignment in the pipeline hook is reached.
```

Author: I prefer my version here.

     The variable will be then overwritten.
   - The `test filtering <#test-filtering>`__ happens *after* a test is instantiated, so the only way to scope a variable assignment is to prefix it with the test class name.
     However, this has some positive side effects:

>      - Doing ``-S valid_systems='*'`` and ``-S valid_prog_environs='*'`` is the equivalent of passing the :option:`--skip-system-check` and :option:`--skip-prgenv-check` options.

Reviewer 2: ```suggestion
     - The command line options ``-S valid_systems='*'`` and ``-S valid_prog_environs='*'`` are equivalent to passing the :option:`--skip-system-check` and :option:`--skip-prgenv-check` options.
```

Author: I'd probably change "Doing" to "Passing".

     - Users could alter the behavior of tests based on tag values that they pass from the command line, by changing the behavior of a test in a post-init hook based on the value of the :attr:`~reframe.core.pipeline.RegressionTest.tags` attribute.

>      - Users could force a test with required variables to run if they set these variables from the command line.

Reviewer 2: ```suggestion
     - Users could force a test with required variables to run only if they set these variables from the command line.
```

Author: I prefer the original text here.

>        For example, the following test could only be run if invoked with ``-S num_tasks=<NUM>``:

Reviewer 2: ```suggestion
       For example, the following test will run only when invoked with ``-S num_tasks=<NUM>``:
```

Author: I want to stress that this is the only way to run it.
